test_python/test_c2.py

```diff
@@ -65,6 +63,42 @@ def ref(X, W):
             reference=ref,
         )
 
+    @given(n=st.integers(1, 128),
+           m=st.integers(1, 128),
+           k=st.integers(1, 128),
+           seed=st.integers(min_value=0, max_value=2**32 - 1),
+           **hu.gcs_gpu_only)
+    @settings(max_examples=2)
+    def test_matmul_tune_and_run(self, n, m, k, seed, gc, dc):
+        matmul = tc.define(MATMUL_LANG, name="matmul")
+
+        mapping_options = matmul.autotune(
+            (n, k), (k, m),
+            generations=1,
+            threads=32,
+            pop_size=2,
+            tuner_min_launch_total_threads=1,
+        )
+
+        X = np.random.rand(m, k).astype(np.float32)
+        W = np.random.rand(k, n).astype(np.float32)
+
+        def ref(X, W):
+            return [np.dot(X, W)]
+
+        op = core.CreateOperator(
+            "TcOp", ["X", "Y"], "out",
+            tcDef=MATMUL_LANG,
+            tcName="matmul",
+            mappingOptions=mapping_options.serialize(),
```

Comment:
you could add a bool parameter here:

`autotune=True` and if its true, then the `genetic_autotuner_caffe2.h` should be called to first tune the kernel on the given tensors. The tuned options can then be cached in memory [this is done for pytorch bindings already in tc_unit.py] and tuning can start from those options to likely yield better kernel OR you can just use those tuned options without re-tuning.

Reply:
@prigoyal , I like this being explicit step instead of tuning manually when the op creates. At least for now. Because then it is clear for users that they need to tune, without tuning it is arbitrary slow.